Refactor ConfigGenerator to replace/set crypt key instead of append

renttek opened the pull request:
Fix the command "setup:config:set --key=..." to replace existing crypt key instead of append it.

### Description
Originally a call so "bin/magento setup:config:set --key='foobar'" did
not replace the crypt key, but append it with a newline.

This does not only "break" handling of values, encryptwed with the old
key, but also don't allow use of new key, because the new key is
composed of the old one, a newline and the new one.


### Fixed Issues (if relevant)
1. magento/magento2#11089: setup:config:set --key append instead of replace

### Manual testing scenarios
1. Install Magento & get sure, that there is a crypt key defined in env.php
2. Call "bin/magento setup:config:set --key=<new key>"
3. New crypt key in env.php should be your new key, instead of "<old>\n<new>"

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

renttek committed:
Refactor ConfigGenerator to replace crypt key instead of append

Originally a call so "bin/magento setup:config:set --key='foobar'" did
not replace the crypt key, but append it with a newline.

This does not only "break" handling of values, encryptwed with the old
key, but also don't allow use of new key, because the new key is
composed of the old one, a newline and the new one.

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -50,15 +51,25 @@ class ConfigGenerator
     /**
+     * @var ConfigDataFactory
+     */
+    protected $configDataFactory;
```
I think this can be made private rather than protected to start with.

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -99,7 +109,7 @@ public function createCryptConfig(array $data)
      */
     public function createSessionConfig(array $data)
     {
-        $configData = new ConfigData(ConfigFilePool::APP_ENV);
```
nice work, good to see this being moved out to a factory.

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -70,15 +81,14 @@ public function createCryptConfig(array $data)
         $currentKey = $this->deploymentConfig->get(ConfigOptionsListConstants::CONFIG_PATH_CRYPT_KEY);
 
-        $configData = new ConfigData(ConfigFilePool::APP_ENV);
+        // TODO: refactor configData to factory in constructor
```
I would suggest leaving the create as it is rather than moving it into the factory. You could always make a `getConfigData` method that checks for `$this->configData` being null and if so calls the factory create.

dmanners replied:
Something like https://github.com/magento/magento2/pull/11152/files#diff-dfb3ee5a7f863458afea75f5524cd2e5R233 might work in this case also.

renttek committed:
Change visibility of configDataFactory-member to private

renttek committed:
Remove unnecessary/old comments

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -50,15 +51,25 @@ class ConfigGenerator
+    public function __construct(
+        Random $random,
+        DeploymentConfig $deploymentConfig,
+        ConfigDataFactory $configDataFactory
```
Optional dependency like http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/#adding-a-constructor-parameter

renttek committed:
Make ConfigDataFactory-parameter optional

dmanners reviewed (changes requested):
Add the optional dependency and I think we are good to go here.

renttek committed:
Refactoring of createCryptConfig

renttek committed:
Introduce CryptKeyGgeneratorInterface

Add CryptKeyGeneratorInterface and a default implementation
"CryptKeyGenerator".
Refactor ConfigGenerator to use CryptKeyGeneratorInterface

renttek committed:
Remove/refactor chained null coalescing operator

to make this code easier to understand, the chained ?? was refactored to
2 simple uses of '.. ?? ..' with some comments

renttek committed:
Make 'Random' optional

Random class is now optional in constructor. If no class is given on
construct, the objectmanager will be used to create an instance

renttek committed:
Fix Unit/Module/ConfigGeneratorTest for Setup

Fixed Magento/Setup/Test/Unit/Module/ConfigGeneratorTest.php

renttek commented:
I made the dependency optional.

Also, I refactored the createCryptConfig method a little bit. (Introduced a new Interface on the way)

renttek committed:
Refactor createCacheHostsConfig

Make it a little more simple

renttek committed:
Reformat some code to make it consistent

renttek committed:
Reformat the big array on the top

renttek committed:
Shorten some function calls in createDbConfig

Put "ConfigOptionsListConstants::CONFIG_PATH_DB_CONNECTION_DEFAULT .
'/'" into a variable to shorten some function calls

renttek committed:
Remove list() to fix some tests

One test reported an undefined "offset 1" for data, that has no port.
Simple solution: remove list() and just work with the array.

renttek committed:
Put "random" parameter back in ConfigGenerator

It has to stay there, so that no older code breaks

renttek committed:
Make "random" required CryptKeyGenerator

dmanners commented on lib/internal/Magento/Framework/Config/Data/ConfigDataFactory.php:
```diff
@@ -0,0 +1,25 @@
+     * @param string $fileKey
+     * @return ConfigData
+     */
+    public function create(string $fileKey)
```
You can update this to use PHP 7's return type declaration if you would like. http://php.net/manual/en/migration70.new-features.php#migration70.new-features.return-type-declarations

renttek committed:
Add random as constructor param in test

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -26,39 +26,49 @@ class ConfigGenerator
-    {
-        $this->random = $random;
+    public function __construct(
+        DeploymentConfig $deploymentConfig,
```
As mentioned we need to keep the `random` parameter here but can mark the property as deprecated.

dmanners commented on setup/src/Magento/Setup/Model/CryptKeyGenerator.php:
```diff
@@ -0,0 +1,52 @@
+     */
+    private $random;
+
+    public function __construct(Random $random = null)
```
Since this is a new class we do not need `random` to be defaulted to null.

renttek commented:
hey @dmanners I pushed the requested changes

dmanners commented on setup/src/Magento/Setup/Model/CryptKeyGenerator.php:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function generate()
```
Again these functions can have the php7 return type deceleration such as `: type` http://php.net/manual/en/migration70.new-features.php#migration70.new-features.return-type-declarations

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -26,39 +27,57 @@ class ConfigGenerator
+        ConfigDataFactory $configDataFactory = null,
+        CryptKeyGeneratorInterface $cryptKeyGenerator = null
+    ) {
         $this->deploymentConfig = $deploymentConfig;
```
Unfortunately we still need to assign `$this->random = $random` here in-case someone is using it in an extension to this class.

renttek replied:
Aah, I thought I added id back.

renttek committed:
Assign deprecated random parameter to property

Has to be done to be backwards compatible

renttek committed:
Add type hints for new interface and impementation

renttek committed:
Add type hint for ConfigDataFactory->create()

dmanners commented:
hey @renttek it looks good to me, there are still a few new functions that could be updated to use the php7 return type declaration but that is not 100% required. If you feel like making that change then please go ahead otherwise I will check the build process and then progress this PR.

renttek committed:
Add some type hints

renttek commented:
Hey,

I added some type hints.
But for the many parts, I can't add type hints to keep BC

dmanners commented on lib/internal/Magento/Framework/Config/Data/ConfigDataFactory.php:
```diff
@@ -0,0 +1,25 @@
+ * Factory for ConfigData
+ * @api
+ */
+class ConfigDataFactory
```
Sorry I missed this before but this factory should either be updated to match the standard defined in http://devdocs.magento.com/guides/v2.2/extension-dev-guide/factories.html or actually it might be possible to use the autogenerated factory here.

orlangur replied:
Autogenerated factories are not allowed in `Magento\Framework` as far as I know. But, the best option is to commit autogenerated factory here.

renttek committed:
Add a empty string return to satisfy return hints

renttek committed:
Add ObjectManager to ConfigDataFactory

ObjectManager is used in ConfigDataFactory to create new instances of
ConfigData

dmanners commented:
So with the updated factory or if it works generated factory then I am happy with this code. If you need help with the factory just drop me a message and I will be happy to help out.

renttek committed:
Refactor test after modifying factory

renttek commented:
Hey, I added the OM to the factory and worked on the tests.
If there are still things to improve, let me know

dmanners commented on setup/src/Magento/Setup/Model/ConfigGenerator.php:
```diff
@@ -227,22 +238,35 @@ public function createModeConfig()
+     *
+     * @return array
+     */
+    private function mapHostData(string $hostData) : array
```
I guess we might need to use php mess dectector tags here so we ignore the error with UnusedPrivateMethod. The method is used in the array_map but the mess dectector does not pick this usage up.

dmanners commented on setup/src/Magento/Setup/Test/Unit/Module/ConfigGeneratorTest.php:
```diff
@@ -18,11 +25,26 @@ class ConfigGeneratorTest extends \PHPUnit\Framework\TestCase
+        $configDataFactoryMock = (new ObjectManager($this))
+            ->getObject(ConfigDataFactory::class, ['objectManager' => $objectManagerMock]);
+
+        $this->configGeneratorObject = new ConfigGenerator($randomMock, $deployConfig, $configDataFactoryMock, $cryptKeyGenerator);
```
might need to split this over a couple of lines so that the static tests do not fail. https://travis-ci.org/magento/magento2/jobs/281732875#L850

renttek committed:
Fix issues of static code checks

renttek committed:
Suppress warnings of long var names in testcases

I Think a shorter variable naming in tests would not only shorten the
code, but also make it harder to understand.

renttek committed:
Remove type hints and @api annotation

- Remove scalar type and return hints
- Inline mapHostData method & get rid of SupressWarning
- Remove @api on new interface & implementation

dmanners commented on lib/internal/Magento/Framework/Config/Data/ConfigDataFactory.php:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * Factory for ConfigData
+ * @api
```
So since it is a patch release we cannot add these tag currently. @renttek Can you check all the classes and remove the `@api` tag if you added it in.

renttek committed:
Some more fixes for static analysis

orlangur commented on setup/src/Magento/Setup/Test/Unit/Module/ConfigGeneratorTest.php:
```diff
@@ -5,24 +5,54 @@
     private $configGeneratorObject;
 
+    /**
+     * @SuppressWarnings(PHPMD.LongVariable)
```
I believe such rule was removed from Magento 2 PHPMD configuration. If it's just Codacy complains, simply ignore this and don't add suppressing annotations. Please remove both `@SuppressWarnings(PHPMD.LongVariable)` introduced.

renttek replied:
Okay, I will remove them. Can such Issues be hidden/ignored on Codacy?

orlangur replied:
Don't know, you can simply ignore it overall. There are some other dumb rules like "never use `else`" :)

Commit:
Remove @SuppressWarning annotations form tests

@SuppresssWarning should not be used

renttek commented:
Hey @orlangur 
I removed the SupressWarnings now

dmanners commented:
@orlangur are you ok with the changes made? If so I will continue to process this.

orlangur reviewed (commented):
Thanks for changes.

orlangur commented:
@dmanners yeah, please proceed, I didn't do a full review, just small note on phpmd suppression.

Looks like "Changes requested" cannot be changed to neutral resolution (and "Approved" is not suitable if I didn't dig into details).

renttek commented:
yeah \o/ first code in core

dmanners commented:
@renttek congrats on the first PR. I have a t-shirt and some stickers for you from the event.

okorshenko commented:
Hi @renttek could you please accept the invite on the GitHub to join magento team?

renttek commented:
Hi @okorshenko I just accepted
